src/mlpack/methods/cf/normalization/z_score_normalization.hpp

```diff
@@ -0,0 +1,162 @@
+/**
+ * @file z_score_normalization.hpp
+ * @author Wenhao Huang
+ *
+ * This class performs z-score normalization on raw ratings.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_CF_NORMALIZATION_Z_SCORE_NORMALIZATION_HPP
+#define MLPACK_METHODS_CF_NORMALIZATION_Z_SCORE_NORMALIZATION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace cf {
+
+/**
+ * This normalization class performs z-score normalization on raw ratings.
+ *
+ * An example of how to use ZScoreNormalization in CF is shown below:
+ *
+ * @code
+ * extern arma::mat data; // (user, item, rating) table
+ * extern arma::Col<size_t> users; // users seeking recommendations
+ * arma::Mat<size_t> recommendations; // Recommendations
+ *
+ * CFType<ZScoreNormalization> cf(data);
+ *
+ * // Generate 10 recommendations for all users.
+ * cf.GetRecommendations(10, recommendations);
+ * @endcode
+ */
+class ZScoreNormalization
+{
+ public:
+  // Empty constructor.
+  ZScoreNormalization() { }
+
+  /**
+   * Normalize the data to zero mean and one standard deviation.
+   *
+   * @param data Input dataset in the form of coordinate list.
+   */
+  void Normalize(arma::mat& data)
+  {
+    mean = arma::mean(data.row(2));
+    stddev = arma::stddev(data.row(2));
+
+    if (std::fabs(stddev) < 1e-14)
+    {
+      Log::Fatal << "Standard deviation of all existing ratings is 0! "
+          << "This may indicate that all existing ratings are the same."
+          << std::endl;
+    }
+
+    data.row(2) = (data.row(2) - mean) / stddev;
+    // The algorithm omits rating of zero. If normalized rating equals zero,
+    // it is set to the smallest positive double value.
+    data.row(2).for_each([](double& x) {
```

Comment:
This should be on a new line: https://github.com/mlpack/mlpack/wiki/DesignGuidelines#brace-placement.

Reply:
@zoq Should this lambda function be written like:
```C++
data.each_col([&](arma::vec& datapoint)
    {
      const size_t item = (size_t) datapoint(1);
      const double rating = datapoint(2);
      itemMean(item) += rating;
      ratingNum(item) += 1;
    });
```

Reply:
Like this:

```c++
data.each_col([&](arma::vec& datapoint)
{
  const size_t item = (size_t) datapoint(1);
  const double rating = datapoint(2);
  itemMean(item) += rating;
  ratingNum(item) += 1;
});
```